### PR TITLE
ENH: initial support for duecredit

### PR DESCRIPTION
**Pull request by yarikoptic**
Only partially addresses #600 
There might be more citations in the code here and probably more to be added within niworkflows .
This use-case pointed (as I left a note in the code) to the chicken&egg problem.  If DOI to be generated only upon the release, duecredit should support fetching citation directly from zenodo (https://github.com/duecredit/duecredit/issues/117)

### CODE REVIEW

File: fmriprep/__init__.py [hunk 1]
```diff
@@ -44,5 +44,19 @@ def _new_version():
     iflogger.setLevel(level)
     if v is None:
         iflogger.warn('afni_vcheck executable not found')
-    return v
```
Comment (effigies) [inline fmriprep/__init__.py]: Guessing this was an accidental removal?

File: fmriprep/__init__.py [hunk 2]
```diff
@@ -44,5 +44,19 @@ def _new_version():
...
 Info.version = staticmethod(_new_version)
+
+from .due import due, Url
```
Comment (effigies) [inline fmriprep/__init__.py]: Could you move the import to the top? Just above warnings seems reasonable.

And I think we need to pull in the `due.py` stub for this to work. Alternately, drop the `.` and add duecredit to the dependencies.

Comment (yarikoptic) [inline fmriprep/__init__.py]: Very sorry @effigies for the rushed up and sloppy work I have done!  I am usually better ;)  Fixed up commit is coming shortly!

File: fmriprep/workflows/epi.py [hunk 3]
```diff
@@ -747,6 +748,11 @@ def _aslist(in_value):
     return workflow
 
 
+@due.cite(
```
Comment (effigies) [inline fmriprep/workflows/epi.py]: `@due.dcite`?

File: fmriprep/workflows/fieldmap/phdiff.py [hunk 4]
```diff
@@ -131,6 +132,11 @@ def _pick1st(inlist):
 # Helper functions
 # ------------------------------------------------------
 
+@due.cite(
```
Comment (effigies) [inline fmriprep/workflows/fieldmap/phdiff.py]: `@due.dcite`?

File: fmriprep/workflows/epi.py (hunk 3 above)
Comment (yarikoptic) [inline fmriprep/workflows/epi.py]: jesus! I need a vacation!!!! pushed the fixed version
sorry again ;)

File: fmriprep/workflows/fieldmap/phdiff.py [hunk 5]
```diff
@@ -131,6 +132,11 @@ def _pick1st(inlist):
 # Helper functions
 # ------------------------------------------------------
 
+@due.dcite(
+    Doi('10.1006/nimg.2001.1054'),
+    description="Conversion of the input phase-difference map into a fieldmap in Hz",
+    tags=['edu']
+)
```
Comment (effigies) [inline fmriprep/workflows/fieldmap/phdiff.py]: Looks like `Function` doesn't play well with decorated functions. I guess upgrading it to a proper interface would be the easiest solution.

Comment (yarikoptic) [inline fmriprep/workflows/fieldmap/phdiff.py]: What is specific problem or Could you please share some easy way to reproduce the problem? (got lost a bit in that circle-ci setup to actually run the test(s))

Comment (effigies) [inline fmriprep/workflows/fieldmap/phdiff.py]: Sure:

```Python
from nipype.pipeline import engine as pe
from nipype.interfaces import utility as niu

@due.dcite(...)
def identfunc(x):
    return x

fnode = pe.Node(niu.Function(function=identfunc), name='fnode')
fnode.inputs.x = 1
fnode.run()
```

The specific problem is that nipype serializes the function as a string, and then reconstitutes it when it's time to run, which doesn't play nicely with decorators.

File: fmriprep/__init__.py [hunk 6]
```diff
@@ -23,6 +23,7 @@
     __description__,
     __longdesc__
 )
+from .due import due, Url
```
Comment (effigies) [inline fmriprep/__init__.py]: Woops. Forgot to import `Doi`.

Comment (yarikoptic) [reply fmriprep/__init__.py]: thanks! pushed the fix.  btw -- feel welcome to push those tiny fixes directly into my fork/branch -- all devs of fmriprep should have permission for that

**Comment by effigies** [comment]
Huh. Apparently I can push to your branch. GitHub's permission structure is weird.

Anyway, @yarikoptic, what's the to-do list here?

**Comment by oesteban** [comment]
>  Apparently I can push to your branch. GitHub's permission structure is weird.

Since a few months ago there's a checkbox you can use to disable permission to repo owners to push to the branch.

**Comment by yarikoptic** [comment]
I guess I would need to check on the laptop first since diff suggests some unrelated changes now... This PR was only for the basic references at that point, so the only Todo was to not cause the disturbance in normal operation of fmriprep ;-)

File: fmriprep/__init__.py [hunk 7]
```diff
@@ -23,8 +23,23 @@
...
 )
+from .due import due, Url, Doi
 
 import warnings
 
 # cmp is not used by fmriprep, so ignore nipype-generated warnings
 warnings.filterwarnings('ignore', r'cmp not installed')
+
+due.cite(
+    # Chicken/egg problem with Zenodo here regarding DOI.  Might need
```
Comment (oesteban) [inline fmriprep/__init__.py]: What is the status regarding this? There is a DOI for the Zenodo entry now, right?

File: fmriprep/interfaces/fmap.py [hunk 8]
```diff
@@ -127,6 +128,50 @@ def _run_interface(self, runtime):
...
+                    Evaluation, NeuroImage 16(1):217-240, 2002. doi:`10.1006/nimg.2001.1054
+                    <http://dx.doi.org/10.1006/nimg.2001.1054>`_.
+
+    """
+    input_spec = PhaseDiff2FieldmapInputSpec
+    output_spec = PhaseDiff2FieldmapOutputSpec
+    references_ = [{
+        'entry': Doi('10.1006/nimg.2001.1054'),
+        'description': "Conversion of the input phase-difference map into a fieldmap in Hz",
+        'tags': ['edu']}]
```
Comment (oesteban) [inline fmriprep/interfaces/fmap.py]: What does this tag mean?

**Comment by oesteban** [comment]
I think we can address this on nipype proper, when we upstream #1024 from here.

When upstreaming, we would make the integration with duecredit and add the feature that @satra proposed (having a flag to stop the visitor going deeper into subworkflows from a given point).

Am I right, @yarikoptic, thinking that you would prefer a general implementation that solves the problem just once within Nipype rather than a custom solution for FMRIPREP (and others)?

WDYT?

Please reopen if you disagree/see it differently.

**Comment by yarikoptic** [comment]
Hi @oesteban ... sure thing IMHO the generic solution within nipype (if possible) would be preferable over adhoc in fmriprep ;-)

**Pull request closed without merging**